src/_posts/2019-10-30-crafting-websites-using-finite-state-machines-part-I.md

```diff
@@ -0,0 +1,403 @@
+
+(defn edit-mode-has-a-valid-selected-id? [state]
+  (if (= "edit" (get-in state [:context :status]))
+    (some (into #{} (map :id (vals (:todos state)))) [(get-in state [:context :selected-id])])
+    true
+    ))
+
+(s/def ::model (s/and
+                 (s/keys :req-un [::context] :opt-un [::todos])
+                 edit-mode-has-a-valid-selected-id?
+                 ))
+
+(comment
+(s/valid? ::model {:context {:status "edit"
+          :filter true
+          :selected-id 1}
+:todos {
+        :1 {:id 1, :text "todo 1", :done? true, :time 1}
+        :2 {:id 2, :text "todo 2", :done? false, :time 2}
+        }}))
+```
+
+### Step 3: Solve the data problem
+
+
+Now we know how we could represent our states as data in the model, let's test drive the different transitions which will al be functions in the controller. Normally I start with a plan, where I will know what I want to test.
+
+The plan of the tests:
+
+ * [ ] controller-should
+ * [ ] initialize-in-list-mode
+ * [ ] check&amp;uncheck
+ * [ ] delete
+ * [ ] toggle-filters
+ * [ ] set a filter
+ * [ ] remove all filters
+ * [ ] set-add-mode
+ * [ ] save-a-new-todo
+ * [ ] set-list-mode
+ * [ ] set-one-for-edit
+ * [ ] save-a-changed-todo
+
+Then we'll write the first one:
+
+```clojure
+(deftest initialize-in-list-mode
+         (is (= {:context {:status "list"} :todos   {}}
+                (init!))))
+```
+
+which will obviously
+
+Test fails! Good... now we write the code to make it pass:
+
+```clojure
+;THE MODEL
+(def model (atom {:context {:status "list"}}))
+
+
+;THE CONTROLLER
+(defn commit! [value atom-to-change]
+  (reset! atom-to-change value))
+
+(defn init! []
+  (-> @model
+    (assoc :todos {})
+    (assoc-in [:context :status] "list")
+    (commit! model)))
+
+
+
+;THE TEST
+(deftest initialize-in-list-mode
+         (is (= {:context {:status "list"} :todos {}}
+                (init!))))
+```
+
+[![TDD](https://img.youtube.com/vi/CvEoXuoAhCg/0.jpg)](https://www.youtube.com/watch?v=CvEoXuoAhCg)
+
+Any refactorings? No, let's move to the second test and so on until we solve the entire data problem, one test at a time, making sure it fails, then making it pass, then refactoring the code. We end up with this:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/controller%20tests.png "Controller tests")
+
+Solve the data problem (TDD the model/controller) Done!
+
+The code is merged here:
+https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/merge_requests/1/diffs
+
+Code:
+
+- [The final code for the controller test](https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/blob/master/test/todomvc/controller_should.cljs)
+- [The final code for the controller](https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/blob/master/src/todomvc/controller.cljs)
+
+Now let's have another look again to see the correspondence between the transitions and the controller functions:
```

Comment:
I think you could expand this section a bit more. This where the whole use of an FSM starts to come into play. Though not sure is coming out correctly. I will expect to see that based on the state of the machine, what i display on the screen changes. But not sure I can see that happening?

Reply:
Yes, the transitions are the controller functions which basically change the model atom. These changes are watched by reagent and the page gets repainted when the model changes. But this is very specific to reagent.

Reply:
Wasn't sure if that was how reagent was working (I only encountered it once briefly before). I think you want to explain, in that case, what it is doing.

Reply:
Added: It is also worth mentioning that we're using the Reagent library, where the model is an atom that is watched for changes. When the atom is changed (in our case by the controller functions, thus the transitions) it will tell the UI to repaint.